### PR TITLE
Test for cmd -c option for agent and controller

### PR DESCRIPTION
**Pull request by ArtiomDivak**
Test for cmd -c option for agent and controller

Test if cmd -c option with invalid conf will
set the status of the agent and controller to failed status.

Related-to: https://github.com/eclipse-bluechi/bluechi/issues/668
Signed-off-by: Artiom Divak <adivak@redhat.com>

### CODE REVIEW

File: tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    restart_with_config_file(f"{config_file_location}/{valid_foo_port_8421_conf_str}", node_foo, bluechi_agent_str)
+    # print(ctrl.exec_run("cat /etc/services ")[1])
+
+    
+    print(node_foo.exec_run("systemctl status bluechi-agent.service ")[1])
+    restart_service(node_bar, bluechi_agent_str)
+    assert check_if_node_disconnected(node_bar)
+    assert check_if_node_connected(NODE_FOO, ctrl, node_foo)
+
+    # print(ctrl.exec_run("systemctl status bluechi-controller.service ")[1])
```
Comment (mkemel) [inline tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py]: Please remove commented code

File: tests/tests/tier0/bluechi-enable-service/test_bluechi_enable_service.py [hunk 2]
```diff
@@ -1,48 +0,0 @@
-# SPDX-License-Identifier: LGPL-2.1-or-later
```
Comment (mkemel) [inline tests/tests/tier0/bluechi-enable-service/test_bluechi_enable_service.py]: Why did you remove these files?

File: tests/tests/tier0/bluechi-enable-service/main.fmf [hunk 3]
```diff
@@ -1,2 +1,2 @@
 summary: Test if a disabled systemd service can be enabled using bluechictl
-id: e3ec3cb3-1010-4205-9b99-507233c6a8f1
+id: e3ec3cb3-1010-4205-9b99-507233c6a8f1
```
Comment (mkemel) [inline tests/tests/tier0/bluechi-enable-service/main.fmf]: Still unnecessary change - add a newline

File: tests/tests/tier0/bluechi-agent-config-c-option/config-files-agent/invalid.conf
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/config-files-agent/invalid.conf:1]: Please keep config files simple and have there only options which you want to override

File: tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        file_location = os.path.join(config_files_ctrl, file)
+        content = read_file(file_location)
+        ctrl.create_file(config_file_location, file_location, content)
+    for file in os.listdir(config_files_agent):
+        file_location = os.path.join(config_files_agent, file)
+        content = read_file(file_location)
+        node_foo.create_file(config_file_location, file_location, content)
+
+    LOGGER.info("Check if every node is up and running without changing the conf with -c")
+    assert ctrl.wait_for_unit_state_to_be(bluechi_controller_str, active_status)
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py]: Not needed, it's already checked in https://github.com/eclipse-bluechi/bluechi/blob/main/tests/bluechi_test/test.py#L74

File: tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    assert node_foo.wait_for_unit_state_to_be(bluechi_agent_str, active_status)
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py]: Not needed, it's already checked in https://github.com/eclipse-bluechi/bluechi/blob/main/tests/bluechi_test/test.py#L93

File: tests/tests/tier0/bluechi-agent-config-c-option/config-files-agent/valid.conf [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+[bluechi-agent]
+#
+# The unique name of this agent.
+# NodeName defaults the system's hostname.
+NodeName="node-bar"
+
+#
+# The IP address that bluechi-agent can use to connect to bluechi manager. It must be a valid IPv4 or IPv6 address.
+# One of the ManagerHost or ManagerAddress options must be set for each bluechi-agent.
+ManagerHost=192.168.1.56
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/config-files-agent/valid.conf]: You can't hardcode IP address of the manager, this IP can be different in different environments

Comment (engelmi) [inline tests/tests/tier0/bluechi-agent-config-c-option/config-files-agent/valid.conf]:
```suggestion
#ManagerHost=192.168.1.56
```
Don't set the `ManagerHost` here. It is determined during test execution and should not be overwritten. This also causes it to fail, currently. 
In more detail:
The agent tries to connect to this host, which probably does not exist on other envs like the CI here. This results in `systemctl stop bluechi-agent` to wait till any timeout will be triggered - in this case the pytest timeout.

**Review by engelmi — COMMENTED** [review]
LGTM, nice work! @ArtiomDivak 

As soon as the suggestion is applied, the integration tests should also pass.

Note: I also created https://github.com/eclipse-bluechi/bluechi/pull/679 to add the serial number to the container names - preventing cascading test failures due to name conflicts.

**Review by mwperina — CHANGES_REQUESTED** [review]
I'd prefer to split this test into several simpler tests:

- Test 1 - check that controller doesn't start if config file passed using `-c` is invalid
- Test 2 - check that agentr doesn't start if config file passed using `-c` is invalid
- Test 3 
    - Check that controller started on default port and agent is connected
    - Restart controller with config file passed using `-c` defining different port, check controller is started and listening on a different port and agent is disconnected
    - Restart agent with config file passed using `-c` defining the same custom port as controller above and check that agent is started and connected to the controller

File: tests/tests/tier0/bluechi-agent-config-c-option/config-files-agent/invalid.conf
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/config-files-agent/invalid.conf:1]: Please remove all unneeded text from this file, so it's clear why it's invalid.
It should contain only following lines:

```
[bluechi-agent]
ManagerPort8420
```

File: tests/tests/tier0/bluechi-agent-config-c-option/config-files-agent/valid.conf
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/config-files-agent/valid.conf:1]: I'd remove this config file and check only the invalid one

File: tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    invalid_conf_str = "invalid.conf"
+    valid_conf_str = "valid.conf"
+
+    # Copying relevant config files into node container
+    for file in os.listdir(config_files_agent):
+        file_location = os.path.join(config_files_agent, file)
+        content = read_file(file_location)
+        node_foo.create_file(config_file_location, file_location, content)
+
+    LOGGER.info("Setting invalid.conf as conf file for foo and checking if failed")
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py]: Please change log level to debug or remove completely

File: tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    restart_with_config_file(os.path.join(config_file_location, invalid_conf_str), node_foo, bluechi_agent_str)
+    assert node_foo.wait_for_unit_state_to_be(bluechi_agent_str, failed_status)
+
+    LOGGER.info("Setting valid.conf as conf file for node_foo and checking if active")
+    restart_with_config_file(os.path.join(config_file_location, valid_conf_str), node_foo, bluechi_agent_str)
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py]: I'd check here only invalid config file and leaving checking valid config file for the port test, because for valid config file we should also check that configuration is applied

File: tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    assert node_foo.wait_for_unit_state_to_be(bluechi_agent_str, active_status)
+
+
+def test_agent_config_c_option(
+        bluechi_test: BluechiTest,
+        bluechi_node_default_config: BluechiNodeConfig, bluechi_ctrl_default_config: BluechiControllerConfig):
+    node_foo_cfg = bluechi_node_default_config.deep_copy()
+    node_foo_cfg.node_name = NODE_FOO
+    node_foo_cfg.manager_port = "8420"
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py]: This is the default port, do we need to override here?

File: tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                       f"{config_file_location}' "
+                       f"/usr/lib/systemd/system/{service}.service")
+    container.exec_run("systemctl daemon-reload")
+    container.exec_run(f"systemctl restart {service}.service")
+
+
+def exec(ctrl: BluechiControllerContainer, nodes: Dict[str, BluechiNodeContainer]):
+
+    node_foo = nodes[NODE_FOO]
+    config_files_agent = "config-files-agent"
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py]: To have the same file structure for all `-c` related tests I'd renamed the directory to `config-files` and have there relavant file names (for example `agent-invalid.conf`)

File: tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+from bluechi_test.util import read_file
+
+LOGGER = logging.getLogger(__name__)
+
+NODE_FOO = "node-foo"
+active_status = "active"
+failed_status = "failed"
+
+
+def restart_with_config_file(config_file_location, container, service):
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/test_agent_config_c_option.py]: This method is used in multiple tests, so it's probably worth to add to `BluechiContainer` class in `bluechi_test/container.py`

File: tests/tests/tier0/bluechi-agent-config-c-option/main.fmf
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-config-c-option/main.fmf:1]: I'd rename the test to `bluechi-agent-cmd-line-c-invalid-config` and update `summary` to contain more clear description

File: tests/tests/tier0/bluechi-ctrl-config-c-option/config-files-ctrl/invalid.conf
Comment (mwperina) [inline tests/tests/tier0/bluechi-ctrl-config-c-option/config-files-ctrl/invalid.conf:1]: Please remove all unneeded text from this file, so it's clear why it's invalid.
It should contain only following lines:

```
[bluechi-controller]
ManagerPort8420
```

File: tests/tests/tier0/bluechi-ctrl-config-c-option/config-files-ctrl/valid.conf
Comment (mwperina) [inline tests/tests/tier0/bluechi-ctrl-config-c-option/config-files-ctrl/valid.conf:1]: I'd remove this config file and check only the invalid one

File: tests/tests/tier0/bluechi-ctrl-config-c-option/main.fmf
Comment (mwperina) [inline tests/tests/tier0/bluechi-ctrl-config-c-option/main.fmf:1]: I'd rename the test to `bluechi-controller-cmd-line-c-invalid-config` and update `summary` to contain more clear description

File: tests/tests/tier0/bluechi-ctrl-config-c-option/test_ctrl_config_c_option.py
Comment (mwperina) [inline tests/tests/tier0/bluechi-ctrl-config-c-option/test_ctrl_config_c_option.py:1]: Similar comments from test_agent_config_c_option can be applied also here

**Review by mwperina — CHANGES_REQUESTED** [review]

File: tests/bluechi_test/container.py [hunk 12]
```diff
@@ -205,6 +205,13 @@ def gather_valgrind_logs(self, data_dir: str) -> None:
         self.extract_valgrind_logs(valgrind_log_path_controller, bluechi_valgrind_log_target_path, data_dir)
         self.extract_valgrind_logs(valgrind_log_path_agent, bluechi_agent_valgrind_log_target_path, data_dir)
 
+    def restart_with_config_file(self, config_file_location, service):
+        self.container.exec_run(f"sed -i '/ExecStart=/c\\ExecStart=/usr/libexec/{service} -c "
```
Comment (mwperina) [inline tests/bluechi_test/container.py]: When this method is moved here shouldn't we use just `self.exec_run()`?

File: tests/tests/tier0/bluechi-agent-cmd-line-c-invalid-config/test_bluechi_agent_cmd_line_c_invalid_config.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+from typing import Dict
+from bluechi_test.test import BluechiTest
+from bluechi_test.container import BluechiControllerContainer, BluechiNodeContainer
+from bluechi_test.config import BluechiControllerConfig, BluechiNodeConfig
+from bluechi_test.util import read_file
+
+LOGGER = logging.getLogger(__name__)
+
+NODE_FOO = "node-foo"
+active_status = "active"
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-cmd-line-c-invalid-config/test_bluechi_agent_cmd_line_c_invalid_config.py]: Unused, please remove

File: tests/tests/tier0/bluechi-agent-cmd-line-c-invalid-config/invalid.conf
Comment (mwperina) [inline tests/tests/tier0/bluechi-agent-cmd-line-c-invalid-config/invalid.conf:1]: Wouldn't it be better to have this file inside `config-files` subdirectory?

File: tests/tests/tier0/bluechi-controller-cmd-line-c-invalid-config/invalid.conf
Comment (mwperina) [inline tests/tests/tier0/bluechi-controller-cmd-line-c-invalid-config/invalid.conf:1]: Wouldn't it be better to have this file inside `config-files` subdirectory?

File: tests/tests/tier0/bluechi-controller-cmd-line-c-invalid-config/test_bluechi_controller_cmd_line_c_invalid_config.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+from typing import Dict
+from bluechi_test.test import BluechiTest
+from bluechi_test.container import BluechiControllerContainer, BluechiNodeContainer
+from bluechi_test.config import BluechiControllerConfig, BluechiNodeConfig
+from bluechi_test.util import read_file
+
+LOGGER = logging.getLogger(__name__)
+
+active_status = "active"
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-controller-cmd-line-c-invalid-config/test_bluechi_controller_cmd_line_c_invalid_config.py]: Unused, please remove

File: tests/tests/tier0/bluechi-controller-cmd-line-c-invalid-config/test_bluechi_controller_cmd_line_c_invalid_config.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    ctrl.create_file(config_file_location, invalid_conf_str, content)
+
+    LOGGER.debug("Setting invalid.conf as conf file for ctrl and checking if failed")
+    ctrl.restart_with_config_file(os.path.join(config_file_location, invalid_conf_str), bluechi_controller_str)
+    assert ctrl.wait_for_unit_state_to_be(bluechi_controller_str, failed_status)
+
+
+def test_agent_config_c_option(
+        bluechi_test: BluechiTest,
+        bluechi_node_default_config: BluechiNodeConfig, bluechi_ctrl_default_config: BluechiControllerConfig):
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-controller-cmd-line-c-invalid-config/test_bluechi_controller_cmd_line_c_invalid_config.py]: BluechiNodeConfig unused, please remove

**Review by mwperina — CHANGES_REQUESTED** [review]

**Review by mwperina — APPROVED** [review]
Great work!
